File: Module_4/Teht_7/7.js
const input = document.getElementById('addressInput');
const btn = document.getElementById('searchBtn');
const results = document.getElementById('results');
const mapDiv = document.getElementById('map');

let map = null;
let routeLayers = [];
let allRouteData = [];

const modeColors = {
    WALK: '#888', BUS: '#007ac9', TRAM: '#00985f',
    RAIL: '#8c4799', SUBWAY: '#ff6319', FERRY: '#00b9e4'
};

// Google encoded polyline decoder
function decodePolyline(encoded) {
    const points = [];
    let index = 0, lat = 0, lng = 0;
    while (index < encoded.length) {
        let b, shift = 0, result = 0;
        do { b = encoded.charCodeAt(index++) - 63; result |= (b & 0x1f) << shift; shift += 5; } while (b >= 0x20);
        lat += (result & 1) ? ~(result >> 1) : (result >> 1);
        shift = 0; result = 0;
        do { b = encoded.charCodeAt(index++) - 63; result |= (b & 0x1f) << shift; shift += 5; } while (b >= 0x20);
        lng += (result & 1) ? ~(result >> 1) : (result >> 1);
        points.push([lat / 1e5, lng / 1e5]);
    }
    return points;
}

function initMap() {
    if (!map) {
        mapDiv.style.display = 'block';
        map = L.map('map').setView([60.17, 24.94], 12);
        L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
            attribution: '&copy; OpenStreetMap contributors'
        }).addTo(map);
    }
}

function clearMapLayers() {
    routeLayers.forEach(layer => map.removeLayer(layer));
    routeLayers = [];
}

function drawRoute(routeIndex) {
    clearMapLayers();
    const node = allRouteData[routeIndex];
    const bounds = L.latLngBounds();

    node.legs.forEach(leg => {
        const coords = decodePolyline(leg.legGeometry.points);
        const color = modeColors[leg.mode] || '#333';
        const line = L.polyline(coords, {
            color, weight: leg.mode === 'WALK' ? 4 : 6,
            opacity: 0.85,
            dashArray: leg.mode === 'WALK' ? '8, 8' : null
        }).addTo(map);
        routeLayers.push(line);
        coords.forEach(c => bounds.extend(c));
    });

    // Origin marker
    const first = node.legs[0];
    const originMarker = L.marker([first.from.lat, first.from.lon]).addTo(map)
        .bindPopup(`<b>Lähtö:</b> ${first.from.name}`);
    routeLayers.push(originMarker);

    // Destination marker
    const last = node.legs[node.legs.length - 1];
    const destMarker = L.marker([last.to.lat, last.to.lon]).addTo(map)
        .bindPopup(`<b>Kohde:</b> ${last.to.name}`);
    routeLayers.push(destMarker);

    map.fitBounds(bounds, { padding: [40, 40] });

    // Highlight active card
    document.querySelectorAll('.route-card').forEach((el, i) => {
        el.classList.toggle('active', i === routeIndex);
    });
}

input.addEventListener('keydown', (e) => {
    if (e.key === 'Enter') searchRoute();
});

async function searchRoute() {
    const address = input.value.trim();
    if (!address) return;

    btn.disabled = true;
    results.innerHTML = '<div class="loading">Haetaan reittejä...</div>';
    allRouteData = [];

    try {
        const resp = await fetch('/api/route', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ address })
        });
        const data = await resp.json();

        if (!resp.ok) {
            results.innerHTML = `<div class="error">${data.error}</div>`;
            return;
        }

        let html = '';

        html += `
            <div class="info-row">
                <div><span class="label">Lähtö</span><br><span class="value">${data.origin.name}</span></div>
                <div><span class="label">Kohde</span><br><span class="value">${data.destination.name}</span></div>
            </div>`;

        const edges = data.routes?.planConnection?.edges || [];
        if (edges.length === 0) {
            html += '<div class="error">Reittejä ei löytynyt</div>';
        }

        edges.forEach((edge, i) => {
            const node = edge.node;
            allRouteData.push(node);
            const startTime = new Date(node.start).toLocaleTimeString('fi-FI', { hour: '2-digit', minute: '2-digit' });
            const endTime = new Date(node.end).toLocaleTimeString('fi-FI', { hour: '2-digit', minute: '2-digit' });
            const totalMin = node.legs.reduce((sum, l) => sum + l.duration, 0);
            const totalMinutes = Math.round(totalMin / 60);

            html += `<div class="route-card" onclick="drawRoute(${i})">`;
            html += `<div class="route-header">
                <h3>Reitti ${i + 1}</h3>
                <span class="route-time">${startTime} → ${endTime} (${totalMinutes} min)</span>
            </div>`;

            node.legs.forEach(leg => {
                const mode = leg.mode;
                const route = leg.trip?.routeShortName || '';
                const label = route ? `${mode} ${route}` : mode;
                const duration = Math.round(leg.duration / 60);

                html += `<div class="leg">
                    <span class="mode-badge mode-${mode}">${label}</span>
                    <div class="leg-details">
                        <div class="leg-stops">${leg.from.name} <span class="arrow">→</span> ${leg.to.name}</div>
                        <div class="leg-duration">${duration} min</div>
                    </div>
                </div>`;
            });

            html += `</div>`;
        });

        results.innerHTML = html;

        // Show map and draw first route
        if (allRouteData.length > 0) {
            initMap();
            drawRoute(0);
        }

    } catch (err) {
        results.innerHTML = `<div class="error">Virhe: ${err.message}</div>`;
    } finally {
        btn.disabled = false;
    }
}

File: Module_4/Teht_7/7.py
from flask import Flask, request, jsonify, send_from_directory
import requests as http_requests
import json
import os

app = Flask(__name__, static_folder=".", static_url_path="")

API_KEY = "79c93aaead7340bbb61f69b496a073a2"
GEOCODE_URL = "https://api.digitransit.fi/geocoding/v1/search"
ROUTING_URL = "https://api.digitransit.fi/routing/v2/hsl/gtfs/v1"

api_headers = {
    "digitransit-subscription-key": API_KEY,
    "Content-Type": "application/json"
}

def geocode(address):
    params = {
        "text": address,
        "size": 1,
        "digitransit-subscription-key": API_KEY
    }
    resp = http_requests.get(GEOCODE_URL, params=params)
    data = resp.json()
    if data["features"]:
        feature = data["features"][0]
        coords = feature["geometry"]["coordinates"]
        name = feature["properties"]["label"]
        return {"lon": coords[0], "lat": coords[1], "name": name}
    return None

def plan_route(from_coords, to_coords):
    query = {
        "query": """
        {
            planConnection(
                origin: {location: {coordinate: {latitude: %f, longitude: %f}}}
                destination: {location: {coordinate: {latitude: %f, longitude: %f}}}
                first: 3
            ) {
                edges {
                    node {
                        start
                        end
                        legs {
                            mode
                            startTime
                            endTime
                            duration
                            from { name lat lon }
                            to { name lat lon }
                            legGeometry { points }
                            trip {
                                routeShortName
                            }
                        }
                    }
                }
            }
        }
        """ % (from_coords["lat"], from_coords["lon"], to_coords["lat"], to_coords["lon"])
    }
    resp = http_requests.post(ROUTING_URL, headers=api_headers, json=query)
    return resp.json()

@app.route("/")
def index():
    return send_from_directory(".", "7.html")

@app.route("/api/route", methods=["POST"])
def get_route():
    data = request.get_json()
    address = data.get("address", "").strip()
    if not address:
        return jsonify({"error": "Osoite puuttuu"}), 400

    origin = geocode(address)
    destination = geocode("Karaportti 2, Espoo")

    if not origin:
        return jsonify({"error": f"Osoitetta '{address}' ei löytynyt!"}), 404
    if not destination:
        return jsonify({"error": "Kohteen 'Karaportti 2' osoitetta ei löytynyt!"}), 404

    result = plan_route(origin, destination)

    return jsonify({
        "origin": origin,
        "destination": destination,
        "routes": result.get("data", {})
    })

if __name__ == "__main__":
    app.run(debug=True, port=5000)